### PR TITLE
feat: Support maximum concurrency of Lambda with SQS as an event source

### PR DESCRIPTION
**Pull request by davidegiunchi**
## Description
Aws recently introduced maximum concurrency of lambda functions when using sqs as an event source ( https://aws.amazon.com/it/blogs/compute/introducing-maximum-concurrency-of-aws-lambda-functions-when-using-amazon-sqs-as-an-event-source/ ), i've added the support of this feature to this module.  
This is my first contribution to this repo, please check the code since i might be wrong.

## Motivation and Context
Limiting lambda concurrency with SQS without setting reserved concurrency.

## Breaking Changes
If you want to set maximum_concurrency_sqs_scaling you need the aws provider >= 4.51.0, because the support for scaling_config has been introduced in 4.51.0: https://github.com/hashicorp/terraform-provider-aws/blob/main/CHANGELOG.md#4510-january-19-2023

## How Has This Been Tested?
- [x] I have updated at least one of the `examples/*` to demonstrate and validate my change(s)
- [ ] I have tested and validated these changes using one or more of the provided `examples/*` projects

- [x] I have executed `pre-commit run -a` on my pull request

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -280,6 +280,14 @@ resource "aws_lambda_event_source_mapping" "this" {
     }
   }
 
+  dynamic "scaling_config" {
+    for_each = try(each.value.maximum_concurrency_sqs_scaling, null) != null ? [true] : []
```
Comment (bryantbiggs) [inline main.tf:284]:
```suggestion
    for_each = try([each.value.maximum_concurrency_sqs_scaling], [])
```

File: main.tf [hunk 2]
```diff
@@ -280,6 +280,14 @@ resource "aws_lambda_event_source_mapping" "this" {
...
+    content {
+      maximum_concurrency = maximum_concurrency_sqs_scaling.value
```
Comment (bryantbiggs) [inline main.tf:286]:
```suggestion
      maximum_concurrency = scaling_config.value.maximum_concurrency
```

File: examples/event-source-mapping/main.tf [hunk 3]
```diff
@@ -24,8 +24,9 @@ module "lambda_function" {
 
   event_source_mapping = {
     sqs = {
-      event_source_arn        = aws_sqs_queue.this.arn
-      function_response_types = ["ReportBatchItemFailures"]
+      event_source_arn                = aws_sqs_queue.this.arn
+      function_response_types         = ["ReportBatchItemFailures"]
+      maximum_concurrency_sqs_scaling = "2"
```
Comment (bryantbiggs) [inline examples/event-source-mapping/main.tf:29]: this would become

```hcl
      scaling_config = {
        maximum_concurrency = 2
      }
```

**Comment by davidegiunchi** [comment]
not stale

**Comment by DStranger** [comment]
@davidegiunchi this is a welcome change! Curious, what is the status of this PR?

**Comment by davidegiunchi** [comment]
@DStranger the PR ha already been checked and corrected by Bryant, we are waiting for the PR to be merged (i think by Anton)

**Comment by jomast** [comment]
is this feature already implemented? https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lambda_event_source_mapping#maximum_concurrency

**Comment by antonbabenko** [comment]
@davidegiunchi Please fix this PR as Bryant suggested, and I will be able to merge it.

**Comment by jonathangueedes** [comment]
@davidegiunchi help us with this please.

**Comment by bryantbiggs** [comment]
let me see if I can push changes - if I can, I can probably get it updated and released - one sec

File: examples/event-source-mapping/main.tf (hunk 3 above)
Comment (bryantbiggs) [reply examples/event-source-mapping/main.tf]: resolved in https://github.com/terraform-aws-modules/terraform-aws-lambda/pull/402/commits/748c6c2d8a1152e020fd84ff3ed61d41d117314a

File: main.tf (hunk 2 above)
Comment (bryantbiggs) [reply main.tf]: resolved in https://github.com/terraform-aws-modules/terraform-aws-lambda/pull/402/commits/748c6c2d8a1152e020fd84ff3ed61d41d117314a

File: main.tf (hunk 1 above)
Comment (bryantbiggs) [reply main.tf]: resolved in https://github.com/terraform-aws-modules/terraform-aws-lambda/pull/402/commits/748c6c2d8a1152e020fd84ff3ed61d41d117314a

File: examples/alias/main.tf [hunk 4]
```diff
@@ -2,7 +2,6 @@ provider "aws" {
   region = "eu-west-1"
 
   # Make it faster by skipping something
-  skip_get_ec2_platforms      = true
```
Comment (bryantbiggs) [inline examples/alias/main.tf:5]: removed due to 
image

File: examples/event-source-mapping/main.tf [hunk 5]
```diff
@@ -20,12 +27,15 @@ module "lambda_function" {
   handler       = "index.lambda_handler"
   runtime       = "python3.8"
 
-  source_path = "${path.module}/../fixtures/python3.8-app1"
+  source_path = "${path.module}/../fixtures/python3.8-app1/index.py"
```
Comment (bryantbiggs) [inline examples/event-source-mapping/main.tf:30]: this was throwing an error around a specific version of python being required. Changing to point directly to the individual file fixed that. I don't believe that affects the intent of this example

File: examples/event-source-mapping/main.tf [hunk 6]
```diff
@@ -216,21 +226,26 @@ resource "aws_kinesis_stream" "this" {
 }
 
 # Amazon MQ
-data "aws_vpc" "default" {
-  default = true
-}
+module "vpc" {
```
Comment (bryantbiggs) [inline examples/event-source-mapping/main.tf:229]: I don't have any default VPCs, so added a bare bones one to make it work for the RabbitMQ requirements

File: examples/event-source-mapping/main.tf [hunk 7]
```diff
@@ -216,21 +226,26 @@ resource "aws_kinesis_stream" "this" {
...
-data "aws_security_group" "default" {
-  vpc_id = data.aws_vpc.default.id
-  name   = "default"
+  enable_nat_gateway = false
 }
 
 resource "aws_mq_broker" "this" {
   broker_name        = random_pet.this.id
   engine_type        = "RabbitMQ"
-  engine_version     = "3.8.11"
```
Comment (bryantbiggs) [inline examples/event-source-mapping/main.tf:231]: this is almost at EOS so updated to latest recommended by AWS

**Comment by antonbabenko** [comment]
This PR is included in version 4.13.0 :tada:

**Pull request merged**